### PR TITLE
Well-formedness check

### LINKED ISSUE #13249: Slow block production on berkeley
For about 2 days, block production on `berkeley` has slowed to about one per hour.

The `whale-1` node has a lot of stake, and wins many slots, but only a few blocks that it's scheduled to produce get produced, in fact.

A possible cause is that there are zkApp transactions that attempt to set a timing with a zero vesting period in an account update, which triggers an assertion in the zkApp transaction logic, seen several times in the whale's logs:
```
{"timestamp":"2023-05-22 03:42:02.503134Z","level":"Error","source":{"module":"Block_producer","location":"File \"src/lib/block_producer/block_producer.ml\", line 276, characters 14-26"},"message":"Error building the diff: $error","metadata":{"error":"Pre_diff_info.Error error: (Unexpected\n \"Assert_failure src/libtransaction_logiczkapp_command_logic.ml:1326:19\")","host":"34.69.16.224","peer_id":"12D3KooWMZ5827QsFKria2YCjgT6iQLsSnGamPkCbBCwv5fedms8","port":10501}}
```
That's confusing because building the diff only looks at the fee payer update, which doesn't mention timing. Apply a diff would look at the other updates, so maybe there's some logic error.

### LINKED ISSUE #13264: Refactor transaction verification
We perform some verification on transactions when admitting them into the transaction pool and when validating transactions in blocks that we receive via gossip. This verification is performed by checking if a transaction is invalid to apply regardless of chain state, and can best be summarized as checking if a transaction is well formed. This includes checks regarding the maximum size of a transaction and other limits.

We should add a new function to the transaction module called `check_well_formedness`. A good signature for this function could be `val check_well_formedness : genesis_constants:Genesis_constants.t -> t -> Well_formedness_error.t list`, which is returning a list of errors that can be logged or returned on an API call.

Here are the functions where this should be called from: https://github.com/MinaProtocol/mina/blob/develop/src/lib/transition_handler/block_sink.ml#L40 https://github.com/MinaProtocol/mina/blob/develop/src/lib/network_pool/transaction_pool.ml#L1035

### PR DESCRIPTION
**Pull request by psteckler**
For user commands, add `check_well_formedness`, which looks for:
- insufficient fee
- too-big zkApps
- zero vesting period in updates to account timing

Call that check when:
- transactions are added to the pool
- gossiped blocks are received
- transactions are submitted via GraphQL

Create unit tests for the zero-vesting-period detection.

Closes #13264. Supersedes #13254.

Probably fixes #13249.

### CODE REVIEW

**Comment by psteckler** [comment]
!ci-build-me

File: src/lib/mina_base/user_command.ml [hunk 1]
```diff
@@ -378,3 +378,53 @@ let valid_size ~genesis_constants = function
...
+          | Set { vesting_period; _ } ->
+              Mina_numbers.Global_slot.(equal zero) vesting_period )
+
+module Well_formedness_error = struct
+  (* syntactically-evident errors such that a user command can never succeed *)
+  type t =
+    | Insufficient_fee
+    | Zero_vesting_period
+    | Zkapp_too_big of
+        (Error.t[@to_yojson fun err -> `String (Error.to_string_hum err)])
```
Comment (mrmr1993) [inline src/lib/mina_base/user_command.ml:399]: Use `Error_json` library here?

**Review by mrmr1993 — APPROVED** [review]
Logic changes LGTM, but see nits

File: src/lib/mina_base/user_command.ml (hunk 1 above)
Comment (psteckler) [reply src/lib/mina_base/user_command.ml:399]: done

File: src/lib/mina_base/user_command.ml [hunk 2]
```diff
@@ -378,3 +378,53 @@ let valid_size ~genesis_constants = function
       Ok ()
   | Zkapp_command zkapp_command ->
       Zkapp_command.valid_size ~genesis_constants zkapp_command
+
+let has_zero_vesting_period = function
+  | Signed_command _ ->
+      false
+  | Zkapp_command (p : Zkapp_command.t) ->
+      Zkapp_command.Call_forest.exists p.account_updates ~f:(fun p ->
```
[the remark this answers is not in the document]
Comment (psteckler) [reply src/lib/mina_base/user_command.ml:386]: done

**Comment by psteckler** [comment]
!ci-build-me

**Comment by psteckler** [comment]
!ci-build-me

**Comment by psteckler** [comment]
!ci-build-me

**Comment by psteckler** [comment]
!ci-build-me

**Comment by psteckler** [comment]
!ci-build-me

**Comment by psteckler** [comment]
!ci-build-me

**Comment by psteckler** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!approved-for-mainnet

**Pull request merged**